File: pc02/pc02_04/python/pc02_04.py
"""
Crypt kicker
pc110204
uva843
"""

import sys


def can_map(cipher_word, plain_word, reverse_key, changed):
    """Check if cipher_word can map to plain_word with current reverse_key."""
    for i in range(len(cipher_word)):
        c = ord(cipher_word[i]) - ord('a')
        p = plain_word[i]

        if reverse_key[c] != '*':
            if reverse_key[c] != p:
                return False
        else:
            for j in range(26):
                if reverse_key[j] == p:
                    return False

            reverse_key[c] = p
            changed[c] = True

    return True


def solve(dictionary, word_index, words, reverse_key):
    """Recursively find a valid mapping."""
    if word_index >= len(words):
        return True

    cipher_word = words[word_index]
    word_len = len(cipher_word)

    for dict_word in dictionary:
        if len(dict_word) != word_len:
            continue

        changed = [False] * 26
        reverse_key_copy = reverse_key[:]

        if can_map(cipher_word, dict_word, reverse_key, changed):
            if solve(dictionary, word_index + 1, words, reverse_key):
                return True

        reverse_key[:] = reverse_key_copy

    return False


def split_words(line):
    """Split line into words."""
    return line.split()


def find_key(dictionary, crypt):
    """Find the mapping key for the cryptogram."""
    words = split_words(crypt)
    reverse_key = ['*'] * 26

    if not solve(dictionary, 0, words, reverse_key):
        reverse_key = ['*'] * 26

    return reverse_key


def kick_crypt(crypt, reverse_key):
    """Decrypt the cryptogram using the reverse key."""
    uncrypt = []

    for c in crypt:
        if c == ' ':
            uncrypt.append(' ')
        else:
            idx = ord(c) - ord('a')
            if reverse_key[idx] == '*':
                uncrypt.append('*')
            else:
                uncrypt.append(reverse_key[idx])

    return ''.join(uncrypt)


def main():
    n = int(input())
    dictionary = []
    for _ in range(n):
        dictionary.append(input().strip())

    for line in sys.stdin:
        crypt = line.rstrip('\n')
        if not crypt:
            continue

        reverse_key = find_key(dictionary, crypt)
        print(kick_crypt(crypt, reverse_key))


if __name__ == '__main__':
    main()

File: pc02/pc02_04/cpp/pc02_04.cpp
/**
 * Crypt kicker
 * pc110204
 * uva843
 */

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

const int ALPHSIZE = 26;

bool canMap(const string& cipherWord, const string& plainWord,
            vector<char>& reverseKey, vector<int>& changed) {
    for (int i = 0; i < cipherWord.length(); i++) {
        int c = cipherWord[i] - 'a';
        char p = plainWord[i];

        if (reverseKey[c] != '*') {
            if (reverseKey[c] != p) {
                return false;
            }
        } else {
            for (int j = 0; j < ALPHSIZE; j++) {
                if (reverseKey[j] == p) {
                    return false;
                }
            }

            reverseKey[c] = p;
            changed[c] = 1;
        }
    }

    return true;
}

bool solve(const vector<string>& dict, int wordIndex, const vector<string>& words,
           vector<char>& reverseKey) {
    if (wordIndex >= words.size()) {
        return true;
    }

    string cipherWord = words[wordIndex];
    int len = cipherWord.length();

    for (const string& dictWord : dict) {
        if (dictWord.length() != len) {
            continue;
        }

        vector<int> changed(ALPHSIZE, 0);

        if (canMap(cipherWord, dictWord, reverseKey, changed)) {
            if (solve(dict, wordIndex + 1, words, reverseKey)) {
                return true;
            }
        }

        for (int j = 0; j < ALPHSIZE; j++) {
            if (changed[j]) {
                reverseKey[j] = '*';
            }
        }
    }

    return false;
}

vector<string> splitWords(const string& line) {
    vector<string> words;
    string word;

    for (char c : line) {
        if (c == ' ') {
            if (!word.empty()) {
                words.push_back(word);
                word.clear();
            }
        } else {
            word += c;
        }
    }

    if (!word.empty()) {
        words.push_back(word);
    }

    return words;
}

void findKey(const vector<string>& dict, const string& crypt,
             vector<char>& reverseKey) {
    vector<string> words = splitWords(crypt);

    if (!solve(dict, 0, words, reverseKey)) {
        fill(reverseKey.begin(), reverseKey.end(), '*');
    }
}

string kickCrypt(const string& crypt, const vector<char>& reverseKey) {
    string uncrypt;

    for (char c : crypt) {
        if (c == ' ') {
            uncrypt += ' ';
        } else {
            int idx = c - 'a';
            if (reverseKey[idx] == '*') {
                uncrypt += '*';
            } else {
                uncrypt += reverseKey[idx];
            }
        }
    }

    return uncrypt;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    cin.ignore();

    vector<string> dict(n);
    for (int i = 0; i < n; i++) {
        getline(cin, dict[i]);
    }

    string crypt;
    while (getline(cin, crypt)) {
        if (crypt.empty()) continue;

        vector<char> reverseKey(ALPHSIZE, '*');

        findKey(dict, crypt, reverseKey);

        cout << kickCrypt(crypt, reverseKey) << '\n';
    }

    return 0;
}
